### PR TITLE
Bradcast functions for audio classes

### LINKED ISSUE #382: Improve broadcasting in dsp.convolve
## General

* pyfar version: 0.5.1
* Python version: 3.10
* Operating System: all
* Did you install pyfar via pip: yes

## Description

`dsp.convolve` requires signals of the same dimension to work. This is often not the case and we could think about adding broadcasting for `signal.cshape` to the function to make it more versatile.

### PR DESCRIPTION
**Pull request by f-brinkmann**
### Changes proposed in this pull request:

Add functions to broadcast signals to specific cshapes and cdims. This can be used to close #382. I thought if it is useful there, it might as well be a separate function. If we decide to keep it in `dsp.utils` we could add functions for concatenating / merging signals here as well.

I would do the tests, if we decide to keep this.

### CODE REVIEW

**Review by sikersten — COMMENTED** [review]
Thanks for adding that, implementation looks good!
I agree that it will help with concatenating signals. I also like the idea of adding a _utils_ submodule, though I would not directly expect it to be a submodule of _dsp_. I mainly associate it with "Signal functionality".
We should discuss if adding a _pyfar.utils_ module one the base level would make sense.

File: pyfar/utils.py [hunk 1]
```diff
@@ -0,0 +1,129 @@
+import pyfar as pf
+import numpy as np
+
+
+def broadcast_cshape(signal, cshape):
+    """
+    Broadcast signal to cshape.
```
Comment (sikersten) [inline pyfar/utils.py:7]:
```suggestion
    Broadcast a signal to a certian cshape.
```

File: pyfar/utils.py [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+    """
+    Broadcast multiple signals to a common cshape.
+
+    The :py:mod:`cshape <pyfar._concepts.audio_classes>` of the signals are
+    broadcasted following the `numpy broadcasting rules
+    <https://numpy.org/doc/stable/user/basics.broadcasting.html>`_
+
+    Parameters
+    ----------
+    signals : tuple of Signal, TimeData, and FrequencyData objects
```
Comment (sikersten) [inline pyfar/utils.py:45]: When building the docs locally, only _TimeData_ is recognized by sphinx and linked. Maybe we can reformulate it so all classes are linked?

File: pyfar/utils.py [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+        The signals to be broadcasted.
+    cshape : tuple, optional
+        The cshape to which the signals are broadcasted. If `chsape` is
```
Comment (sikersten) [inline pyfar/utils.py:48]:
```suggestion
        The cshape to which the signals are broadcasted. If `cshape` is
```

File: pyfar/utils.py [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+            raise TypeError("All input data must be pyfar audio objects")
+
+    if cshape is None:
+        cshape = np.broadcast_shapes(*[s.cshape for s in signals])
+    return [broadcast_cshape(s, cshape) for s in signals]
+
+
+def broadcast_cdim(signal, cdim):
+    """
+    Broadcast signal to cdim.
```
Comment (sikersten) [inline pyfar/utils.py:69]:
```suggestion
    Broadcast a signal to a certain cdim.
```

File: pyfar/utils.py [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+        The cshape to which `signal` is broadcasted.
+
+    Returns
+    -------
+    signal : Signal, TimeData, FrequencyData
+        The broadcasted input signal
+    """
+
+    if not isinstance(signal, (pf.Signal, pf.TimeData, pf.FrequencyData)):
+        raise TypeError("All input data must be pyfar audio objects")
```
Comment (sikersten) [inline pyfar/utils.py:27]:
```suggestion
        raise TypeError("Input data must be a pyfar audio object.")
```

File: pyfar/utils.py [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+        The cdim to which `signal` is broadcasted.
+
+    Returns
+    -------
+    signal : Signal, TimeData, FrequencyData
+        The broadcasted input signal
+    """
+
+    if not isinstance(signal, (pf.Signal, pf.TimeData, pf.FrequencyData)):
+        raise TypeError("All input data must be pyfar audio objects")
```
Comment (sikersten) [inline pyfar/utils.py:90]:
```suggestion
        raise TypeError("Input data must be a pyfar audio object.")
```

File: pyfar/utils.py [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+    ----------
+    signal : Signal, TimeData, FrequencyData
+        The signal to be broadcasted.
+    cshape : tuple
+        The cshape to which `signal` is broadcasted.
+
+    Returns
+    -------
+    signal : Signal, TimeData, FrequencyData
+        The broadcasted input signal
```
Comment (sikersten) [inline pyfar/utils.py:23]: Should we mention explicitly that (in contrast to numpy) a copy of the signal is returned?

File: pyfar/utils.py [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+    return [broadcast_cshape(s, cshape) for s in signals]
+
+
+def broadcast_cdim(signal, cdim):
+    """
+    Broadcast signal to cdim.
+
+    The channel dimension (cdim) is the length of the
+    :py:mod:`cshape <pyfar._concepts.audio_classes>` of the signal. The signal
+    is broadcasted to `cdim` by prepending ``cdim - len(signal.cshape)``
```
Comment (sikersten) [inline pyfar/utils.py:73]: Could it be useful to append instead of prepending?

**Review by sikersten — CHANGES_REQUESTED** [review]
Only minor comments regarding docs and one question regarding _broadcast_cdim_ :)

File: pyfar/utils.py (hunk 1 above)
Comment (f-brinkmann) [reply pyfar/utils.py:7]: Done...

File: pyfar/utils.py (hunk 7 above)
Comment (f-brinkmann) [reply pyfar/utils.py:23]: Yes, good idea

File: pyfar/utils.py (hunk 5 above)
Comment (f-brinkmann) [reply pyfar/utils.py:27]: Yes, better

File: pyfar/utils.py (hunk 2 above)
Comment (f-brinkmann) [reply pyfar/utils.py:45]: yes, linked them manually

File: pyfar/utils.py (hunk 4 above)
Comment (f-brinkmann) [reply pyfar/utils.py:69]: done :)

File: pyfar/utils.py (hunk 8 above)
Comment (f-brinkmann) [reply pyfar/utils.py:73]: This is following the standard numpy rules for broadcasting. Thats why I did not add an option for this.

File: pyfar/utils.py (hunk 6 above)
Comment (f-brinkmann) [reply pyfar/utils.py:90]: done

File: pyfar/utils.py (hunk 2 above)
Comment (sikersten) [reply pyfar/utils.py]: Locally, FrequencyData is linked now and Signal still is not... but I also think this is not very important

File: pyfar/utils.py (hunk 7 above)
Comment (sikersten) [reply pyfar/utils.py]: Could also be added to the other functions :)

File: pyfar/utils.py (hunk 8 above)
Comment (sikersten) [reply pyfar/utils.py:73]: ok, thanks for clarifying

File: pyfar/dsp/__init__.py [hunk 9]
```diff
@@ -37,6 +37,7 @@
 __all__ = [
     'fft',
     'filter',
+    'utils',
```
Comment (ahms5) [inline pyfar/dsp/__init__.py:40]: Why do we need to add it here, since utils is in the parent dir?

File: pyfar/utils.py [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+    The channel dimension (cdim) is the length of the
+    :py:mod:`cshape <pyfar._concepts.audio_classes>` of the signal. The signals
+    are broadcasted to `cdim` by prepending ``cdim - len(signal.cshape)``
+    dimensions.
+
+    Parameters
+    ----------
+    signals : tuple of Signal, TimeData, FrequencyData
+        The signals to be broadcasted in a tuple.
+    cdim : int
```
Comment (ahms5) [inline pyfar/utils.py:114]:
```suggestion
    cdim : int, optional
```

File: pyfar/dsp/__init__.py (hunk 9 above)
Comment (f-brinkmann) [reply pyfar/dsp/__init__.py:40]: Thanks for noting. That was a leftover

File: pyfar/utils.py (hunk 10 above)
Comment (f-brinkmann) [reply pyfar/utils.py:114]: yes!

File: pyfar/utils.py (hunk 7 above)
Comment (f-brinkmann) [reply pyfar/utils.py]: True, added to all functions

File: pyfar/utils.py (hunk 2 above)
Comment (f-brinkmann) [reply pyfar/utils.py]: Works now: https://pyfar.readthedocs.io/en/feature-bradcast_audio_classes/modules/pyfar.utils.html

Comment (sikersten) [reply pyfar/utils.py]: Nice! Did not work with sphinx 5.1.1 which I had installed locally, sorry!

File: pyfar/utils.py [hunk 11]
```diff
@@ -0,0 +1,129 @@
...
+        The signals to be broadcasted in a tuple.
+    cshape : tuple, optional
+        The cshape to which the signals are broadcasted. If `cshape` is
+        ``None`` it is determined from the cshapes of the input signals using
+        ``numpy.broadcast_shapes``. The default is ``None``.
+
+    Returns
+    -------
+    signals : tuple of Signal, TimeData, FrequencyData
+        The broadcasted copies input signals in a tuple
```
Comment (sikersten) [inline pyfar/utils.py:55]:
```suggestion
        The broadcasted copies of the input signals in a tuple.
```

File: pyfar/utils.py [hunk 12]
```diff
@@ -0,0 +1,129 @@
...
+    signals : tuple of Signal, TimeData, FrequencyData
+        The signals to be broadcasted in a tuple.
+    cdim : int, optional
+        The cdim to which `signal` is broadcasted. If `cdim` is ``None`` the
+        signals are broadcasted to the largest `cdim`. The default is ``None``.
+
+    Returns
+    -------
+    signals : tuple of Signal, TimeData, FrequencyData
+        The broadcasted copies input signals in a tuple
```
Comment (sikersten) [inline pyfar/utils.py:121]:
```suggestion
        The broadcasted copies of the input signals in a tuple.
```

**Review by sikersten — APPROVED** [review]

**Pull request merged**